tests/failpoints/cases/test_async_fetch.rs

```diff
@@ -103,3 +102,121 @@ fn test_node_cache_compact_with_one_node_down() {
     }
     panic!("cluster is not compacted after inserting 200 entries.");
 }
+
+// Test the case that async fetch is performed well while the peer is removed.
+#[test]
+fn test_node_async_fetch_remove_peer() {
+    let count = 5;
+    let mut cluster = new_node_cluster(0, count);
+    cluster.pd_client.disable_default_operator();
+
+    cluster.cfg.raft_store.raft_log_gc_count_limit = 100000;
+    cluster.cfg.raft_store.raft_log_gc_threshold = 50;
+    cluster.cfg.raft_store.raft_log_gc_size_limit = ReadableSize::mb(20);
+    cluster.cfg.raft_store.raft_log_gc_tick_interval = ReadableDuration::millis(100);
+    cluster.cfg.raft_store.raft_log_reserve_max_ticks = 2;
+    cluster.cfg.raft_store.raft_entry_cache_life_time = ReadableDuration::millis(100);
+    cluster.run();
+
+    cluster.must_put(b"k1", b"v1");
+    cluster.must_transfer_leader(1, new_peer(1, 1));
+
+    // cause log lag and trigger async fetch
+    cluster.stop_node(5);
+    for i in 1..60 {
+        let k = i.to_string().into_bytes();
+        let v = k.clone();
+        cluster.must_put(&k, &v);
+    }
+    fail::cfg("worker_async_fetch_raft_log", "pause").unwrap();
+    cluster.run_node(5).unwrap();
+
+    // make sure destroy_peer and on_entries_fetched are called in one batch.
+    cluster.must_transfer_leader(1, new_peer(2, 2));
+    fail::cfg("worker_gc_raft_log", "pause").unwrap();
+    cluster.pd_client.must_remove_peer(1, new_peer(1, 1));
+    fail::cfg("pause_on_peer_collect_message", "pause").unwrap();
+    fail::remove("worker_gc_raft_log");
+    sleep_ms(10);
+    fail::remove("worker_async_fetch_raft_log");
+    sleep_ms(10);
```

Comment:
Every subtle.

Reply:
make it be prone to follow the order that destroy peer is handled before on entries fetched